proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+These new variables hold values which are a result of some arbitrary function invocation on the existing variables
+in the invocation context.
+
+InputOutputShaper is especially useful when combined with PromptNodes. Aside from simply renaming variables to match
+the templates of PromptNodes, we can also use InputOutputShaper to add new variables to the invocation context. Often
+these new variables are the result of some arbitrary function invocation on the existing variables in the invocation.
+
+The original idea for InputOutputShaper was to support the use case of expanding the query string variable to a list of
+strings and the list size matching the size of the documents list. This is useful when we want to use the same query string
+as a question to pose to all the documents in the documents list.
+
+The expected outcome of using InputOutputShaper is that we can easily add new variables to the invocation context
+so they can match the prompt templates of PromptNodes. Multiple InputOutputShaper components can be used in a pipeline to
+modify the invocation context as needed.
+
+
+# Detailed design
+
+The InputOutputShaper component is primarily used in pipelines and configured via a YAML file. The YAML component
+definition consists of the params block:
+
+```yaml
+            components:
+            - name: shaper
+              params:
+                inputs:
+                   query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                   documents:
+                      func: contract_docs
+                      params:
+                        docs: documents
+                        delimiter: " "
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+```
+
+The params block consists of the inputs list. Each item in the inputs list is a dictionary with the key being the
+invocation context variable that we want to modify.
+
+In the example above, we have two items in the inputs list. The first item is a dictionary with the key `query` and the
+second item is a dictionary with the key `documents`.
+
+For the query variable, we want to invoke the function `expand` and store the result in the variable `questions`.
+The `expand` function takes two keyword parameters: `expand_target` and `size`. The `expand_target` parameter is the
+name of the variable in the invocation context that we want to expand. The `size`parameter is a result of the `len`
+function invocation on the variable `documents`.
+
+For the documents variable, we want to invoke the function `contract_docs` and store the result in the same variable.
```

Comment:
`contract_docs` might be confusing here. What about calling this function `concat` and adding an `in_place` parameter? That parameter could make all the functions more flexible. No need to introduce `contract_docs` and `contract_...` if we need the concatenation in future not only for documents. 
The `expand` function (and any other future functions) could also get an `in_place` parameter.

Reply:
`concat` sounds great but what is the objective of `in_place`?

Reply:
The `in_place` or `inplace` parameter would make setting `output` parameter obsolete if there is no renaming of the input. We could reuse the concat function (and other functions) and have `in_place` set to `True` by default. 

Right now, `contract_docs` can't be easily reused let's say for answers because the default output is docs. My suggestion is to rename the function to just `concat` and have a `in_place` or `inplace` set to `True` by default, which should have the effect that the output is the same as the input. `documents` if it's used on `documents` and `answers` if it's used on `answers`. 

I got this thought because I am working with pandas right now. Something like `drop_duplicates(inplace=True) `https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html
Whether that's a good idea for Haystack depends on whether many functions will be reused for different inputs (documents, answers, etc.) and whether we always rename the output or sometimes have the same output name as the input.

Reply:
Aha gotcha, this was already in the implementation, but I forgot to mention it here. If there is an output parameter specified - then it is used, otherwise, the value of the invocation is assigned to the input variable. I'll add a section on "Ommiting output parameter"